src/store.c

```diff
@@ -339,144 +374,255 @@ static void writer_close(
+    writer_close(&store, 0);
+  fail_encode_a:
+    coder_close(&coder_a);
+    indexer_free(indexer_b);
+  fail_indexer_b_alloc:
+    indexer_free(indexer_a);
+  fail_indexer_a_alloc:
   fail_open:
     free(vals);
   fail_vals:
     return false;
 }
 
+static struct vals *vals_from_index(struct index* index)
+{
+    struct vals *vals =
+        calloc(1, sizeof(*vals) + sizeof(vals->data[0]) * index->len);
+
+    vals->len = index->len;
+
+    for (size_t i = 0; i < index->len; ++i) {
+        vals->data[i] = index->data[i].key;
+    }
+
+    return vals;
+}
+
+static bool merge_with_config(
+    struct encoder* coder,
+    struct rill_store** list,
+    size_t list_len,
+    enum rill_col col)
+{
+  struct rill_kv kvs[list_len];
+
+  struct decoder decoders[list_len];
+
+  size_t it_len = 0;
+  for (size_t i = 0; i < list_len; ++i) {
+    if (!list[i]) continue;
+    decoders[it_len] = store_decoder(list[i], col);
+    it_len++;
+  }
+  assert(it_len);
+
+  for (size_t i = 0; i < it_len; ++i) {
+    if (!(coder_decode(&decoders[i], &kvs[i]))) { /* goto fail_coder; // TODO FIXME */ }
+  }
+
+  struct rill_kv prev = {0};
+
+  while (it_len > 0) { /* while there are valid entries of rill db files */
+    size_t target = 0;
+
+    for (size_t i = 1; i < it_len; ++i) {
+      if (rill_kv_cmp(&kvs[i], &kvs[target]) < 0)
+        target = i;
+    }
+
+    struct rill_kv *kv = &kvs[target];
+    struct decoder *decoder = &decoders[target];
+
+    if (rill_likely(rill_kv_nil(&prev) || rill_kv_cmp(&prev, kv) < 0)) {
+      if (!coder_encode(coder, kv)) { /* goto fail_coder; // TODO FIXME */ }
+      prev = *kv;
+    }
+
+    if (!coder_decode(decoder, kv)) { /* goto fail_coder; // TODO FIXME */ }
+    if (rill_unlikely(rill_kv_nil(kv))) {
+      memmove(kvs + target,
+              kvs + target + 1,
+              (it_len - target - 1) * sizeof(kvs[0]));
+      memmove(decoders + target,
+              decoders + target + 1,
+              (it_len - target - 1) * sizeof(decoders[0]));
+      it_len--;
+    }
+  }
+
+  return true;
+}
+
 bool rill_store_merge(
         const char *file,
         rill_ts_t ts, size_t quant,
         struct rill_store **list, size_t list_len)
 {
     assert(list_len > 1);
 
-    size_t keys = 0;
     size_t pairs = 0;
     struct vals *vals = NULL;
+    struct vals *invert_vals = NULL;
 
-    struct decoder decoders[list_len];
-    struct rill_kv kvs[list_len];
-
-    size_t it_len = 0;
     for (size_t i = 0; i < list_len; ++i) {
         if (!list[i]) continue;
         vma_will_need(list[i]);
 
-        struct vals *ret = vals_merge(vals, list[i]->vals);
-        if (ret) { vals = ret; } else { goto fail_vals; }
+        struct vals *merge_from_index = vals_from_index(list[i]->index_b);
+        struct vals *invert_merge_from_index = vals_from_index(list[i]->index_a);
+        struct vals *ret = vals_merge(vals, merge_from_index);
+        struct vals *iret = vals_merge(invert_vals, invert_merge_from_index);
```

Comment:
Instead of loading the values in a list then merging them in, you could simply merge them in from the index directly. This will greatly simplify the logic and avoid memory leaks on error which I think are there right now.

Reply:
Fixed, shot myself in the foot in the process, but got some duck tape, so eh :D.